README.md
# influxdb_store

[![main](https://github.com/knorth55/influxdb_store/actions/workflows/main.yml/badge.svg)](https://github.com/knorth55/influxdb_store/actions/workflows/main.yml)
[![linter](https://github.com/knorth55/influxdb_store/actions/workflows/linter.yaml/badge.svg)](https://github.com/knorth55/influxdb_store/actions/workflows/linter.yaml)

ROS package for influxdb store

![influxdb_grafana](./media/influxdb_grafana.png)

## Sample

```bash
roslaunch influxdb_store influxdb_logger.launch
```

## Logger

### `joint_states_logger.py`

Logger for `sensor_msgs/JointStates`

#### Subscribing topic

- `~input` (`sensor_msgs/JointStates`)

Joint states topic name

#### Parameters

- `~host` (default: `localhost`)

Influxdb host address

- `~port` (default: `8086`)

Influxdb port number

- `~database` (default: `test`)

Influxdb database name

### `map_transform_logger.py`

Logger for transfrom relative to `map_frame`

#### Parameters

- `~host` (default: `localhost`)

Influxdb host address

- `~port` (default: `8086`)

Influxdb port number

- `~database` (default: `test`)

Influxdb database name

- `~map_frame_id` (default: `map`)

Map frame id

- `~update_rate` (default: `0.5`)

Update rate

### `battery_states_logger.py`

Logger for `pr2_msgs/BatteryServer2`

#### Subscribing topic

- `~input` (`pr2_msgs/BatteryServer2`)

Battery topic name

#### Parameters

- `~host` (default: `localhost`)

Influxdb host address

- `~port` (default: `8086`)

Influxdb port number

- `~database` (default: `test`)

Influxdb database name

### `network_states_logger.py`

Logger for network information published by [`jsk_network_tools/network_status.py`](https://github.com/jsk-ros-pkg/jsk_common/tree/master/jsk_network_tools)

#### Subscribing topic

- `~input/receive` (`std_msgs/Float32`)

receiving bps topic name

- `~input/transmit` (`std_msgs/Float32`)

Transmitting bps topic name

#### Parameters

- `~host` (default: `localhost`)

Influxdb host address

- `~port` (default: `8086`)

Influxdb port number

- `~database` (default: `test`)

Influxdb database name

## For JSK PR2 users

### PR1012

```bash
rossetip
rossetmaster pr1012
roslaunch influxdb_store pr2_influxdb_logger.launch robot_name:=pr1012
```

### PR1040

```bash
rossetip
rossetmaster pr1040
roslaunch influxdb_store pr2_influxdb_logger.launch robot_name:=pr1040
```

### Fetch15

```bash
rossetip
rossetmaster fetch15
roslaunch influxdb_store fetch_influxdb_logger.launch robot_name:=fetch15
```

### Fetch1075

```bash
rossetip
rossetmaster fetch1075
roslaunch influxdb_store fetch_influxdb_logger.launch robot_name:=fetch1075
```

### Systemctl services

Please see [jsk_database](https://github.com/knorth55/jsk_database).


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package influxdb_store
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.2 (2021-10-06)
------------------
* Update README.md
* add system scripts
* check joint_states for connection test
* try for 2 minutes
* add sysmtemd scripts
* add required and respawn
* Contributors: Shingo Kitagawa

0.0.1 (2021-08-18)
------------------
* add baxter_influxdb_logger.launch
* disable jscpd linter
* use pytz
* add LookupException in transform_logger
* add respawn true
* fix transform_logger
* fix typo
* update TransformLogger to log in 30 hz
* updata transform logger
* add base_transform_logger in influxdb_logger.launch
* add base_transform_logger
* add TransformLogger
* update readme
* add theta for rotation
* update fetch interface name
* add robot_name in sample launch
* add network_states_logger in influxdb_logger.launch
* add network_states_logger.py
* remove unused line
* update joint_states_logger.py
* flake8
* markdown lint
* enable linter
* update readme
* Merge pull request `#1 <https://github.com/knorth55/influxdb_store/issues/1>`_ from knorth55/use-github-actions
* update README.md
* add github actions
* add python-influxdb
* use threading in joint_states_logger
* use time_precision=ms
* change queue_size
* add fetch_influxdb_logger.launch
* fix pr2_battery_states_logger.py
* add exception in map_transform_logger.py
* split influxdb_logger.launch and pr2_battery_logger.launch
* rename battery_states_logger to pr2_battery_states_logger
* update battery_states_logger
* update readme
* update README
* refactor map_transform_logger
* rafactor joint_states_logger
* add map_transform_logger
* add database arg
* add media
* update readme
* use timestamp_to_influxdb_time
* add python package
* add pr2_influxdb_logger.launch
* add BatteryStatesLogger
* add launch
* add joint_states_logger.py
* add README.md
* add ros package
* Initial commit
* Contributors: Shingo Kitagawa


node_scripts/map_transform_logger.py
#!/usr/bin/env python

import influxdb
import numpy as np
import rospy
import yaml

import tf2_ros

from influxdb_store.utils import timestamp_to_influxdb_time
from tf2_msgs.srv import FrameGraph


class MapTransformLogger(object):
    def __init__(self):
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)

        rospy.wait_for_service("~tf2_frames")
        self.get_frames = rospy.ServiceProxy("~tf2_frames", FrameGraph)

        host = rospy.get_param('~host', 'localhost')
        port = rospy.get_param('~port', 8086)
        database = rospy.get_param('~database', 'test')
        self.client = influxdb.InfluxDBClient(
            host=host, port=port, database=database)
        self.client.create_database(database)

        self.map_frame_id = rospy.get_param('~map_frame_id', 'map')
        self.update_rate = rospy.Duration(
            rospy.get_param('~update_rate', 0.5))
        self.timer = rospy.Timer(self.update_rate, self._cb)

    def _cb(self, event):
        if event.last_real:
            timestamp = event.last_real
        else:
            timestamp = event.current_real - self.update_rate
        time = timestamp_to_influxdb_time(timestamp)
        try:
            graph = yaml.load(self.get_frames().frame_yaml)
        except Exception as e:
            rospy.logerr(e)
            return

        query = []
        for child_id, info in graph.items():
            try:
                transform_stamped = self.tf_buffer.lookup_transform(
                    self.map_frame_id, child_id, timestamp)
            except tf2_ros.ExtrapolationException as e:
                rospy.logerr_throttle(
                    60.0, 'tf2_ros.ExtrapolationException: {}'.format(e))
                continue
            except tf2_ros.ConnectivityException as e:
                rospy.logerr_throttle(
                    60.0, 'tf2_ros.ConnectivityException: {}'.format(e))
                continue
            translation = transform_stamped.transform.translation
            rotation = transform_stamped.transform.rotation
            theta = 2 * np.arctan(
                np.linalg.norm(
                    [rotation.x, rotation.y, rotation.z]) / rotation.w)
            query.append({
                "measurement": "map_transform",
                "tags": {
                    "frame_id": transform_stamped.header.frame_id,
                    "child_frame_id": transform_stamped.child_frame_id
                },
                "time": time,
                "fields": {
                    "translation.x": translation.x,
                    "translation.y": translation.y,
                    "translation.z": translation.z,
                    "rotation.x": rotation.x,
                    "rotation.y": rotation.y,
                    "rotation.z": rotation.z,
                    "rotation.w": rotation.w,
                    "rotation.theta": theta,
                }
            })
        if len(query) > 0:
            self.client.write_points(query, time_precision='ms')


if __name__ == '__main__':
    rospy.init_node('map_transform_logger')
    logger = MapTransformLogger()
    rospy.spin()
